### DISCUSSION TITLE
About midi plugins, and Werner's old MusE-2

### DISCUSSION THREAD

**Discussion by terminator356** [muse-sequencer/muse#1174] [Team Posts]
A new LV2 midi processing plugin was announced today:
[https://github.com/sjaehn/BSchaffl](url)

It reminded me and got me thinking.
Werner's original MusE-2, now called muse-evolution in the attic branch,
 was different from our MusE-2 which we carried on from MusE-1.
(Some of the components used there are found in our tree, simply because I imported
 the much more updated Deicsonze synth from there.)

[ One of things he did was bring the midi ports and devices list out of the shadows.
He eliminated 'ports', with just unlimited 'devices'. I have mentioned plans to do that here.
Then, he made a new track type: Midi Device.
If I recall there were two track types: Midi Input Device and Midi Output Device.
Thus midi devices were made front and centre part of the track list.
Think of them as the logical midi equivalent of Audio Input and Output tracks.
He also made each midi track be able to deliver to multiple output devices.
 where we have just one output device at a time. That's something I've been
 trying to plan for soooo long now. I won't get into the difficulties too much -
 the engine can handle it, but now more information must be presented to the user,
 for example they must choose an instrument on each device that a track routes to,
 and since that feature is also available on the tracks themselves, one has
 multiple instruments to select. ]

Now here's what I wanted to mention: He added a **midi plugin rack**
 to each Midi Track and Midi Device Track.
Thus midi tracks and devices could process midi right in the rack.
This was before LV2 came along, so the plugins were our own.
So I also had on the back burner some plans to add such a rack to our strips.
This new midi LV2 plugin looks like a good reason.
[ Edit ] In fact there were **two** such **audio** or **midi** racks per strip: Input and output plugins,
 as found in many other DAWs. I hope you can see where I am trying to go with those configurable
 racks as seen in the code - to be completed, I thought about it in detail again the other day.

**Comment by kybos** [comment]
Interesting.
Sorry I don't know the history, but why was the current MusE forked from Werner's MusE-1? I always thought it was the continuation of Werner's latest abandoned version...

**Comment by terminator356** [comment]
1 was much more stable, and by the time I came along and submitted
 a bunch of fixes, 2 wasn't very stable and abandoned.
Robert will recall the day I first contacted him with the fixes, and he told me of
 this version 2 which I didn't even know about.
Some time later, we talked about whether to work on 2 or stick with 1.
We decided to continue with 1 because it was already stable and usable.
To get 2 ready would have taken much longer.
There were some apparent conceptual difficulties in 2, some of which would be hard to fix.
For example there were controller lanes separate from the tracks but underneath them.
Unfortunately they could not be placed inside movable parts, so the only option was to
 have a feature 'move controller points with parts'. This concept may sound like a good idea
 but actually it is broken in practice. For example once you move some controllers like that,
 they get mixed in with all the other points in the area, and there's no way to reselect them
 and move them further. Here in this MusE, we have the opportunity to put the audio controllers
 inside parts like the midi controllers. But this has not been done yet.

**Comment by kybos** [comment]
OK, thanks for the explanation!
